File: Data Structures & Algorithms/max-area-of-island/submission-0.java
class Solution {
    
    public static int [][] directions = {{1,0},{-1,0},{0,1},{0,-1}};
    public int maxAreaOfIsland(int[][] grid) {
        int n=grid.length;
        int m=grid[0].length;

        int Area=0;

        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j]==1)
                {
                    Area=Math.max(Area,dfs(grid,i,j));
                }
            }
        }
        return Area;
    }

    private int dfs(int[][] grid, int r, int c)
    {
        if(r<0 || c<0 || r>=grid.length || c>=grid[0].length || grid[r][c]==0)
            return 0;


        grid[r][c]=0;
        int res=1;
        for(int[] row : directions)
        {
            res+=dfs(grid,r+row[0],c+row[1]);
        }
        return res;
    }
}

File: Data Structures & Algorithms/swim-in-rising-water/submission-0.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        '''
            0,0 , 1
            0,1,2
        
        '''
        res = 0
        n = len(grid)

        if n == 1:
            return grid[0][0]
        min_heap = [(grid[0][0],0,0)]
        
        dp = [[float("inf") for _ in range(n)] for _ in range(n)]

        Dir = [[0,1],[0,-1],[1,0],[-1,0]]

        while min_heap:
            elevation,i,j= heapq.heappop(min_heap)
            if i == n-1 and j==n-1:
                return dp[n-1][n-1]
            
            for dr,dc in Dir:
                nr,nc = dr + i, dc + j

                if nr < 0 or nc < 0 or nr >= n or  nc >= n:
                    continue
                
                if max(elevation,grid[nr][nc]) < dp[nr][nc]:
                    dp[nr][nc] = max(elevation,grid[nr][nc]) 
                    heapq.heappush(min_heap,(dp[nr][nc],nr,nc))
         
        return dp[n-1][n-1]
            

